### PR TITLE
Allow to select multiple database

### PR DESCRIPTION
**Pull request by lecorguille**
Hi,

It's a long story of intership students.
So, just a little PR to allow multiple selection of database.

I didn't test this modification since a long time. I will try to test it tomorrow 

Next to that, maybe you remember a talk at Baltimore about a visualization for Blast :/
It's finalized ... but a svn repot. I will try to resurrect the project :)

### CODE REVIEW

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 1]
```diff
@@ -352,7 +352,8 @@
     <token name="@THREADS@">-num_threads "\${GALAXY_SLOTS:-8}"</token>
     <token name="@BLAST_DB_SUBJECT@">
 #if $db_opts.db_opts_selector == "db":
-  -db "${db_opts.database.fields.path}"
+  #set $dbl = " ".join(str( $db_opts.db_type.database.fields.path ).split( ',' ))
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: That should probably not have the ``.db_type`` bit in it which is causing the TravisCI failures:

```
NotFound: cannot find 'db_type' while searching for 'db_opts.db_type.database.fields.path'
```

We can probably do this in one line without needing to set ``$dbl`` too...

**Comment by lecorguille** [comment]
I will check that tomorrow

**Comment by lecorguille** [comment]
So ... 
I'm really sorry but the logs are not clear for me. Can you give me a clue to patch my PR ?

**Comment by peterjc** [comment]
Travis CI Python 2.6 and Galaxy dev looks like a Galaxy problem: https://github.com/galaxyproject/galaxy/issues/1571

Travis CI Python 2.7 and Galaxy dev looks like an SQLite locking issue, re-running the tests now...

**Comment by peterjc** [comment]
The TravisCI failure for Python 2.6 and Galaxy dev turned out to be a planemo dependency issue, see discussion on https://github.com/galaxyproject/planemo/pull/405

The TravisCI failure for Python 2.7 and Galaxy dev did go away on a re-test, consistent with it being a stochastic SQLite locking issue.

So that's good news for your pull request :)

**Comment by peterjc** [comment]
Do you have any thoughts on the resulting user-interface change? Did your users have any trouble accidentally selecting multiple databases?

**Comment by peterjc** [comment]
What I would like to see before releasing this enhancement is at least one test using this. Are you familiar with how this works - or is this something I should handle?

*Notes on potential test cases:* We have two test databases ``four_human_proteins`` and ``four_human_proteins_taxid`` defined in ``test-data/blastdb_p.loc`` (see https://github.com/peterjc/galaxy_blast/blob/master/test-data/blastdb_p.loc ), but they are essentially the same database (with and without taxonomy information) so it would be a bit of an artificial example. Maybe we can add another tiny nucleotide database in ``test-data/blastdb.loc``? See also https://github.com/peterjc/galaxy_blast/blob/master/tools/ncbi_blast_plus/update_test_files.sh

**Comment by peterjc** [comment]
How should I credit this work in the history log? "Gildas Le Corguillé and his students?"

**Comment by peterjc** [comment]
Screenshots with the changes applied, the UI change might flummox some users but seems quite straightforward:

screenshot 2016-01-26 17 07 52

When clicked on with the mouse:

screenshot 2016-01-26 17 08 03

Multiple selections happen by clicking:

screenshot 2016-01-26 17 08 39

When done, the selected databases are shown like this (with easy to use remove crosses):

screenshot 2016-01-26 17 08 49

When you only want one database, this is more complicated - especially if you want to change your first selection. Here's the current view for comparison:

screenshot 2016-01-26 17 12 46

So yes, this UI change could confuse some users, but on balance seems OK to me.

**Comment by lecorguille** [comment]
Hi,

TEST:
No problem, I will build a test using two database as soon as possible.

CREDIT:
Even if I didn't do that for credit, I'm not again some publicity to accelerate the merge of my PR on Galaxy Project :stuck_out_tongue_closed_eyes:. 
So, let say Gildas Le Corguillé and Emma Prudent

UI:
To be honest, we didn't test this feature with a lot of users. So, I can say if it's confusing in its new shape.

**Comment by lecorguille** [comment]
A problem with the dev version of Galaxy?

**Comment by lecorguille** [comment]
Hello,

Do I have to re-launch Travis?
If yes ... how?

**Comment by peterjc** [comment]
I suspect only repository admins can request the TravisCI rebuild, which I have just done.

Hopefully with https://github.com/peterjc/galaxy_blast/commit/b979c46cc6f48de4e83039ee46d058aeb52c6460 this should just work - I doubt your changes are failing.

**Comment by bgruening** [comment]
Nice feature @lecorguille !

**Comment by peterjc** [comment]
I stumbled into another Galaxy problem holding up the automated tests https://github.com/galaxyproject/galaxy/issues/1882 - it turns out Galaxy's dev branch has now officially dropped Python 2.6 support as per https://github.com/galaxyproject/galaxy/issues/1596

I'm working on updating my TravisCI configuration to exclude this.

**Comment by peterjc** [comment]
I finally have Galaxy tests running again on TraivsCI after all the changes from Galaxy v16.01, and started work on your PR.

I've got a branch at https://github.com/peterjc/galaxy_blast/tree/multi which I may still rebase / force update, which includes a squashed commit of your key changes and (currently) an alternative test case. I hope to apply this to the master tomorrow.

**Comment by lecorguille** [comment]
I'm learning every day of your best practice.

I will try to apply those mecanims to our project.

Cheers

2016-03-08 18:54 GMT+01:00 Peter Cock <notifications@github.com>:

**Comment by peterjc** [comment]
@bgruening and others have been doing great work on the IUC and devteam repositories taking the automated testing mechanism with TravisCI and planemo even further - I need to look at that for my own repositories.

**Comment by peterjc** [comment]
@lecorguille Your work is now included on the master branch as https://github.com/peterjc/galaxy_blast/commit/b6c6be06d4d1cad8aa76aaf0a101cb5d21cc7fc6 and acknowledged in the README file. Thank you :)

This is now on the Test Tool Shed if you would like to test it from there https://testtoolshed.g2.bx.psu.edu/view/devteam/ncbi_blast_plus/ - Assuming now problems, I will release this to the Main Tool Shed.

P.S. GitHub hasn't linked your commits to your GitHub account yet - you'd need to add your sb-roscoff.fr email address to your GitHub account.

**Comment by lecorguille** [comment]
Great
Thanks a lot for your interest in this PR

2016-03-09 12:19 GMT+01:00 Peter Cock <notifications@github.com>:

**Pull request closed without merging**